src/server/master_server/master_metadata_service_impl.cc

```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
   return server::MetadataManager::GetInstance();
 }
 
+inline server::ChunkServerManager& MasterMetadataServiceImpl
+    ::chunk_server_manager() {
+  return server::ChunkServerManager::GetInstance();
+}
+
 grpc::Status MasterMetadataServiceImpl::HandleFileCreation(
     const protos::grpc::OpenFileRequest* request,
     protos::grpc::OpenFileReply* reply) {
+  // Step 1. Create file metadata
   const std::string& filename(request->filename());
   google::protobuf::util::Status status(
       metadata_manager()->CreateFileMetadata(filename));
-  return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+  if (!status.ok()) {
+    return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+  }
+
+  // Step 2. Create the first file chunk for this file
+  google::protobuf::util::StatusOr<std::string> chunk_or(
+      metadata_manager()->CreateChunkHandle(filename,0));
+  if (!chunk_or.ok()) {
+    return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+  }
+  const std::string& chunk(chunk_or.ValueOrDie());
+
+  // Step 3. Allocate chunk servers
+  const ushort num_of_chunk_replica(3);
+  // TODO(someone): make this number configurable via config.yml. Probably not
+  // a priority right now
+  auto chunk_server_locations(
+      chunk_server_manager().AllocateChunkServer(chunk, num_of_chunk_replica));
+
+  // Prepare the InitChunkFileReply
+  *reply->mutable_metadata() = FileChunkMetadata();
+  reply->mutable_metadata()->set_chunk_handle(chunk);
+  reply->mutable_metadata()->set_version(0);
+
+  // Step 4. Coordinate with chunk servers to initialize the file chunk 
+  for(auto chunk_server_location : chunk_server_locations) {
+    const std::string server_address(chunk_server_location.server_hostname() +
+        std::to_string(chunk_server_location.server_port()));
+    auto chunk_server_channel(grpc::CreateChannel(server_address,
+        grpc::InsecureChannelCredentials()));
+    // Register this chunk server Rpc client if not existed
+    RegisterChunkServerRpcClient(server_address, chunk_server_channel);
```

Comment:
yikes. in my PR, I deprecated this function to only support `server_address`. prob gonna have to resolve this merge conflict at some point

Reply:
Yeah, I think can I understand why you want to support server_address only. If the server_address already exists, we should not need to create the channel. But this becomes a tricky and challenging thing to do with parallel hash map due to the extra operation that is involved, so I have to keep the channel variable.

Reply:
not quite sure about what you meant by parallel hashmap? FYI, my new way of doing it is instead of having <server_name, client_instance>, i have < server_address, cient_instance> so we can also add more servers in that were not in the initial config file